core/include/vecmem/containers/impl/device_vector.ipp

```diff
@@ -225,6 +227,91 @@ VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::push_back(
     return index;
 }
 
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::bulk_append(size_type n)
+    -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+    static_assert(std::is_default_constructible<TYPE>::value,
+                  "Type `T` in `device_vector<T>::bulk_append` is not a "
+                  "default-constructible type.");
```

Comment:
What happens when we instantiate `vecmem::device_vector` in a non-default-constructible type? 🤔 Even if the user doesn't try to use this function, I think (based on what we saw in #275) that this assertion will activate.

I fear that we can only do this using SFINAE...

To make sure, please add a unit test with a type that's not default constructible. 🤔

Reply:
This is a very fun edge case in C++ that was actually resolved in C++23 with P2593R1. Before that, `static_assert(false)` was _always_ ill-formed, even in templates that are not instantiated. After P2593R1, `static_assert(false)` is only ill-formed if the template is actually instantiated, and can also be hidden in `if constexpr`. For dependent `static_asserts` like this one, it's actually fine: the error is only emitted when the template is actually instantiated.